File: shared/hub/Diagnostics.cs
using System.Text.Json;

namespace NarrationUtilsHub;

/// <summary>Writes append-only startup and runtime diagnostics to JSON Lines.</summary>
public sealed class SessionDiagnostics
{
    private readonly object _lock = new();
    public string SessionDir { get; }
    public string Path { get; }
    public string Identifier { get; }

    public SessionDiagnostics(string sessionDir)
    {
        SessionDir = sessionDir;
        Directory.CreateDirectory(SessionDir);
        Path = System.IO.Path.Combine(SessionDir, "host.log");
        Identifier = new DirectoryInfo(SessionDir).Name;
    }

    public void Event(string name, IReadOnlyDictionary<string, object?>? details = null)
    {
        var record = new Dictionary<string, object?>
        {
            ["time"] = DateTimeOffset.Now.ToString("yyyy-MM-ddTHH:mm:sszzz"),
            ["event"] = name,
            ["session"] = Identifier,
        };
        if (details != null) foreach (var (key, value) in details) record[key] = value;
        var line = JsonSerializer.Serialize(record, new JsonSerializerOptions { Encoder = System.Text.Encodings.Web.JavaScriptEncoder.UnsafeRelaxedJsonEscaping });
        lock (_lock) File.AppendAllText(Path, line + "\n");
    }

    public void Exception(string name, Exception exception) =>
        Event(name, new Dictionary<string, object?> { ["error"] = exception.Message, ["traceback"] = exception.ToString() });
}

File: shared/server/diagnostics.py
"""Append-only startup/runtime diagnostics, ported 1:1 from shared/hub/Diagnostics.cs."""

import json
import os
import threading
from datetime import datetime, timezone
from pathlib import Path


class SessionDiagnostics:
    def __init__(self, session_dir: str):
        self.session_dir = session_dir
        os.makedirs(session_dir, exist_ok=True)
        self.path = Path(session_dir) / "host.log"
        self.identifier = Path(session_dir).name
        self._lock = threading.Lock()

    def event(self, name: str, details: dict | None = None) -> None:
        record = {
            "time": datetime.now(timezone.utc).astimezone().isoformat(timespec="seconds"),
            "event": name,
            "session": self.identifier,
        }
        if details:
            record.update(details)
        line = json.dumps(record, ensure_ascii=False)
        with self._lock:
            with open(self.path, "a", encoding="utf-8") as handle:
                handle.write(line + "\n")

    def exception(self, name: str, exc: Exception) -> None:
        self.event(name, {"error": str(exc), "traceback": repr(exc)})
